doc/cookbook/db-mysql-basics/MysqlBasics.lhs

```diff
@@ -0,0 +1,189 @@
+
+- Our API will be exposed on localhost:8080
+
+## REST actions available
+
+*Get all people*
+
+```
+/people    GET
+```
+
+*Get person by ID*
+
+```
+/people/:id    GET
+```
+
+*Insert a new person*
+
+```
+/people    POST
+
+{
+  "name": "NewName",
+  "age": 24
+}
+```
+
+*Delete a person*
+
+```
+/people/:id    DELETE
+```
+
+## Other notes
+
+At the time of writing this issue may occur when running `stack build`:
+
+```
+setup: Missing dependencies on foreign libraries:
+* Missing (or bad) C libraries: ssl, crypto
+```
+
+This can be fixed by adding the following lines to your `stack.yaml`:
+
+```
+extra-include-dirs:
+- /usr/local/opt/openssl/include
+extra-lib-dirs:
+- /usr/local/opt/openssl/lib
+```
+
+## Implementation: Main.hs
+
+Let's jump in:
+
+```haskell
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE QuasiQuotes                #-}
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TypeFamilies               #-}
+{-# LANGUAGE TypeOperators              #-}
+module MysqlBasics where
+
+import           Control.Monad.IO.Class       (liftIO)
+import           Control.Monad.Logger         (NoLoggingT (..))
+import           Control.Monad.Trans.Reader   (runReaderT)
+import           Control.Monad.Trans.Resource (ResourceT, runResourceT)
+import           Data.Aeson
+import           Data.Int                     (Int64 (..))
+import           Data.Text                    (Text)
+import qualified Data.Text                    as T
+import qualified Data.Text.IO                 as T
+import           Database.Persist
+import           Database.Persist.MySQL       (ConnectInfo (..),
+                                               SqlBackend (..),
+                                               defaultConnectInfo, fromSqlKey, runMigration,
+                                               runSqlPool, toSqlKey, withMySQLConn)
+import           Database.Persist.Sql         (SqlPersistT, runSqlConn)
+import           Database.Persist.TH          (mkMigrate, mkPersist,
+                                               persistLowerCase, share,
+                                               sqlSettings)
+import           Database.Persist.Types       (PersistValue(PersistInt64))
+
+import           GHC.Generics
+import           Network.Wai
+import           Network.Wai.Handler.Warp
+import           Servant
+import           Servant.API
+import           System.Environment           (getArgs)
+
+share [mkPersist sqlSettings, mkMigrate "migrateAll"] [persistLowerCase|
+Person json
+    Id   Int Primary Unique
+    name Text
+    age  Text
+    deriving Eq Show Generic
+|]
+
+data Error = Error404 | Error503 deriving (Generic, Show)
+
+data ApiResponse =
```

Comment:
Servant already has a notion of "successful response OR error response". The `Handler` monad has an instance of `MonadExcept ServerError`, which means you can use `throwErrror` with values of type `ServerError` whenever appropriate, and just otherwise return what would be the successful response directly, without going through a type like this. Roughly something like:

``` haskell
type API = "person" :> Capture "name" String :> Get '[JSON] Person

getPerson :: String -> Handler Person
getPerson name = do
  mperson <- findPerson name
  case mperson of
    Nothing -> throwError ...
    Just p    -> return p
```

You can even define handlers in an `ExceptT AppErrorType IO` monad instead, so as to throw nice, clean Haskell values as errors instead of using `ServerError` which is pretty much an arbitrary HTTP response. That really is an idiomatic way to handle errors in web apps, judging from all the servant code I've seen.

Reply:
Ah - cool! Had no idea that was possible. Will make that change 😄 thanks!